### PR TITLE
Wengert/barchart

### PR DESCRIPTION
**Pull request by phantomxc**
This is a very large refactor to the BarChart component.

- Fixes the re-rendering issue.
- Combines positive and negative values into a single SVG
- Removes the spans for labels
- Now accepts xAxis and yAxis elements allowing you to do axis math from d3.
- More style control.
- A few more things...

### CODE REVIEW

File: docs/components/BarChartDocs.js [hunk 1]
```diff
@@ -127,97 +111,93 @@ const BarChartDocs = React.createClass({
...
-        <p>A style object that can be passed to style the tooltip.</p>
+        <h5>tooltipFormat<label>Function</label></h5>
+        <p>A function that is called to format the value being displayed in the tooltip.</p>
 
         <h5>width<label>Number</label></h5>
         <p>Width of the SVG; bars will determine their individual width accordingly.</p>
         <p>Default: 500</p>
 
+        <h5>xAxis<label>Element</label></h5>
+        <p>A element representing the xAxis. You are incharge of transforming to the correct location.</p>
```
Comment (jmophoto) [inline docs/components/BarChartDocs.js]: Space between in and charge.

File: src/components/BarChart.js [hunk 2]
```diff
@@ -1,299 +1,351 @@
...
   getInitialState () {
     return {
-      hovering: false
+      hoveringObj: {},
+      clickedData: {}
     };
   },
 
   shouldComponentUpdate (nextProps, nextState) {
     return !_isEqual(nextProps, this.props) || !_isEqual(nextState, this.state);
```
Comment (jmophoto) [inline src/components/BarChart.js]: Do you need to omit the functions like you did in the bar?

**Comment by jmophoto** [comment]
Besides one question and a wording change, this looks good to go. 🍪

File: src/components/BarChart.js (hunk 2 above)
Comment (phantomxc) [inline src/components/BarChart.js]: Since that is more the public API, I don't think we can guarantee that those functions aren't changing? That's my thinking anyways...thoughts?

File: docs/components/BarChartDocs.js [hunk 3]
```diff
@@ -89,9 +48,27 @@ const BarChartDocs = React.createClass({
     });
   },
 
+
   render () {
     const styles = this.styles();
 
+    const ticks = chartData.map(d => {
```
Comment (paniclater) [inline docs/components/BarChartDocs.js]: Are we cool with one letter function parameters? I am, especially in one liners like this.

File: src/components/BarChart.js [hunk 4]
```diff
@@ -1,299 +1,351 @@
...
-    primaryColor: React.PropTypes.string,
-    value: React.PropTypes.number.isRequired,
+    radius: React.PropTypes.number,
+    value: React.PropTypes.number,
     width: React.PropTypes.number,
-    x: React.PropTypes.number.isRequired,
-    y: React.PropTypes.number.isRequired
-  },
-
-  mixins: [SetIntervalMixin],
```
Comment (paniclater) [inline src/components/BarChart.js]: <3 thank you for removing mixins <3

File: docs/components/BarChartDocs.js (hunk 3 above)
Comment (cerinman) [inline docs/components/BarChartDocs.js]: In this case I'm cool with them 👍

File: src/components/BarChart.js (hunk 4 above)
Comment (cerinman) [inline src/components/BarChart.js]: Yes, I was very surprised when I saw this.  Very glad to see it removed.  👍

File: docs/components/BarChartDocs.js (hunk 3 above)
Comment (jmophoto) [inline docs/components/BarChartDocs.js]: Kind of a standard in D3 to use `d`. All about the branding.

File: src/components/BarChart.js (hunk 2 above)
Comment (cerinman) [inline src/components/BarChart.js]: Hmm interesting question.  I can't imagine a scenario where I would want changing a function prop to trigger a render.

File: src/components/BarChart.js [hunk 5]
```diff
@@ -1,299 +1,351 @@
...
 
-  componentWillReceiveProps (nextProps) {
-    if (nextProps.value !== this.props.value) {
-      this.setState({
-        milliseconds: 0
-      });
-    }
+  shouldComponentUpdate (nextProps) {
+    // check the props that aren't functions
+    return !_isEqual(_omit(nextProps, _functions(nextProps)), _omit(this.props, _functions(this.props)));
```
Comment (paniclater) [inline src/components/BarChart.js]: I think this comment is unneccesary, _omit and _functions are pretty clear.

File: src/components/BarChart.js [hunk 6]
```diff
@@ -1,299 +1,351 @@
...
   },
 
-  tick () {
-    this.setState({
-      milliseconds: this.state.milliseconds + 10
-    });
+  componentDidUpdate () {
+    if (this.props.hovering && this.props.animateOnHover) {
```
Comment (paniclater) [inline src/components/BarChart.js]: nice job grouping side effects in componetDidUpdate.

File: src/components/BarChart.js [hunk 7]
```diff
@@ -1,299 +1,351 @@
...
+    return `translate(${sx},${sy}) scale(1, ${factor})`;
   },
 
-  _handleMouseOut () {
-    this.setState({
-      hovering: false
-    });
-
-    this.props.onMouseOut();
+  _drawPath ({ x, y, width, height, value, radius }) {
```
Comment (paniclater) [inline src/components/BarChart.js]: i think this is great, makes it very clear what's going on in the path.

File: src/components/BarChart.js [hunk 8]
```diff
@@ -1,299 +1,351 @@
...
+          .duration(100)
+          .attr('transform', 'scale(1)');
+    }
   },
 
-  _handleMouseOver (label, value, x, y) {
-    const animateDuration = 500;
+  _getTransformWithScale (factor) {
+    const centerX = 0;
+    const centerY = (this.props.value > 0) ? this.props.y + this.props.height : this.props.y;
```
Comment (paniclater) [inline src/components/BarChart.js]: why do you only add height if y is positive?

File: src/components/BarChart.js [hunk 9]
```diff
@@ -1,299 +1,351 @@
...
-        onMouseOut={this._handleMouseOut}
-        onMouseOver={this._handleMouseOver.bind(null, this.props.label, this.props.value, this.props.x, y)}
-        rx={3}
-        ry={3}
-        style={style}
-        width={this.props.width}
-        x={this.props.x}
-        y={y}
+      <path
+        d={this._drawPath({
```
Comment (paniclater) [inline src/components/BarChart.js]: so much better than seeing a bunch of logic directly in the d attribute.

File: src/components/BarChart.js (hunk 8 above)
Comment (phantomxc) [inline src/components/BarChart.js]: It has to do with the direction of the transform that needs to happen for the intro animation and the bounce animation.

ezgif com-video-to-gif 15

Comment (paniclater) [inline src/components/BarChart.js]: I don't fully understand, but I believe the gif 💯

File: src/components/BarChart.js (hunk 4 above)
Comment (paniclater) [inline src/components/BarChart.js]: Mixins were the main thing preventing me from substituting React for Preact in the open source. I think if we could get rid of all the mixins we could probably substantially lower our bundle size with Preact.

**Comment by phantomxc** [comment]
I believe I fixed all the concerned items.

**Comment by cerinman** [comment]
:shipit:

**Comment by phantomxc** [comment]
Could I get another peak into the last two commits? Found some issues.

File: src/components/BarChart.js (hunk 4 above)
Comment (lukeschunk) [inline src/components/BarChart.js]: oh also switching to ES6 classes might help with the preact switch. just sayin

File: src/components/BarChart.js [hunk 10]
```diff
@@ -168,7 +169,7 @@ const BarChart = React.createClass({
 
       transform = `translate(${x}, ${y})`;
     } else {
-      transform = 'translate(-100, -100)';
+      transform = 'translate(-1000, -1000)';
```
Comment (lukeschunk) [inline src/components/BarChart.js]: why the changes here?

Comment (phantomxc) [inline src/components/BarChart.js]: This hides the tooltip further out of the SVG viewbox. When on a small screen, there were situations where it should show like a pixel of the tooltip.

**Comment by lukeschunk** [comment]
:shipit:

**Comment by phantomxc** [comment]
Third times a charm right? I didn't account for a proper domain when all the negative numbers were the same. 
༼ノಠل͟ಠ༽ノ ︵ ┻━┻

**Comment by phantomxc** [comment]
Can someone take a look at my last commit?

**Comment by phantomxc** [comment]
image

**Comment by lukeschunk** [comment]
:shipit:

**Comment by lukeschunk** [comment]
:shipit:
